[![MIT](https://img.shields.io/pypi/l/Escher.svg)](https://github.com/erolkavvas/escher/blob/master/LICENSE)

# Metabolic Network Classifiers

Metabolic Network Classifiers (MNCs) are flux balance analysis-based genome classifiers that achieve prediction accuracy on par with state-of-the-art machine learning approaches while enabling a mechanistic interpretation of the genotype-phenotype map. MNCs thus provide a FBA framework for microbial genome-wide association studies (GWAS) (preprint SOON). Inference and computation of MNCs utilize cobrascape (https://github.com/erolkavvas/cobrascape), which wraps around COBRApy package (https://github.com/opencobra/cobrapy) to model a population of strain-specific genome-scale models (GEMs) at allelic resolution.

![cobrafig](/MNC\_overview.png?raw=true)

## Generating Metabolic Network Classifiers
Run `01_sample_mncs.py` and `02_eval_mncs.py`
1. Run `$ python 01_sample_mncs.py -f INPUT_DIR -s NUM_SAMPLES -o MNC_DIR [optional parameters...]`
	- Generates a MNC samples. Requires performing flux variability analysis (popFVA) which is very computationally expensive. Recommend >1000 samples
2. Run `$ python 02_eval_mncs.py -f MNC_DIR [--testset --bicthresh]`
	- Optimize the MNC for the training set or testset.

An ensemble of sampled MNCs can be generated by running steps 2 and 3 multiple times for the same output directory. The input parameters must be consistent for the same output directory. This is helpful because generating samples takes a long time so the simulations are likely to be interrupted at some point.

## Analyzing Metabolic Network Classifiers
Once a large number of MNCs has been generated (~1000+), those with $\deltaBIC<10$ can be analyzed to study the genetic and metabolic basis for the GWAS dataset.
1. See `analyze_mncs.ipynb`

## Input Data
- The following files are to be placed in INPUT_DIR. The files should be named exactly as described below.
    - `MODEL_FILE.json` (**REQUIRED**). A genome-scale model in json format
    - `X_ALLELES_FILE.csv` (**REQUIRED**). A strain allele matrix with shape (strains, alleles). Alleles (columns) should have the same name as the gene ids in the genome-scale model (with \_num appended to specify the allele id) 
    - `Y_PHENOTYPES_FILE.csv` (**REQUIRED**). A strain phenotypes matrix with shape (strains, phenotypes)
    - `GENE_LIST_FILE.csv` (_OPTIONAL_). A pandas series or dataframe with index specifying the list of genes to limit the modeled alleles to. Highly recommended <200 genes. otherwise sample deeply  
-  `cobra_model/gene_to_pathways.json`. JSON object describing mapping of genes to pathways for the organism. Required for pathway enrichments.
- `gene_to_name.json`. JSON object describing mapping of gene ids to names. OPTIONAL.

Strain allele matrix and strain phenotypes matrix from (https://rdcu.be/9rHj) and cobra model from (https://rdcu.be/bG6JO).

## Installation
	$ git clone https://github.com/erolkavvas/metabolic-network-classifiers.git


